File: prototype/java/PrototypeFactory.java
package prototype.java;

public class PrototypeFactory {
    public FruitPrototype fp;
    public VegetablePrototype vp;

    public PrototypeFactory() {
        fp = new FruitPrototype();
        vp = new VegetablePrototype();
    }

    public Prototype clone1() {
        return fp;
    }

    public Prototype clone2() {
        return vp;
    }
}

File: prototype/golang/prototypefactory.go
package main

type PrototypeFactory struct {
}

func (pf *PrototypeFactory) createThing(thing string) Prototype {
	if thing == "fruit" {
		fp := FruitPrototype{}
		return fp.clone()
	}

	if thing == "vegetable" {
		vp := VegetablePrototype{}
		return vp.clone()
	}

	return nil
}
